content_scripts/mode.coffee

```diff
@@ -0,0 +1,210 @@
+#   Optional.  Define a badge is the badge is constant.  Otherwise, do not define a badge and override the
+#   chooseBadge method instead.  Or, if the mode *never* shows a badge, then do neither.
+#
+# keydown:
+# keypress:
+# keyup:
+#   Key handlers.  Optional: provide these as required.  The default is to continue bubbling all key events.
+#
+# Additional handlers associated with the mode can be added by using the push method.  For example, if a mode
+# responds to "focus" events, then push an additional handler:
+#   @push
+#     "focus": (event) => ....
+# Any such additional handlers are removed when the mode exits.
+#
+# New mode types are created by inheriting from Mode or one of its sub-classes.  Some generic cub-classes are
+# provided below:
+#
+#   SingletonMode: ensures that at most one instance of the mode is active at any one time.
+#   ConstrainedMode: exits the mode if the an indicated element loses the focus.
+#   ExitOnEscapeMode: exits the mode on escape.
+#   StateMode: tracks the current Vimium state in @enabled and @passKeys.
+#
+# To install and existing mode, use:
+#   myMode = new MyMode()
+#
+# To remove a mode, use:
+#   myMode.exit() # externally triggered.
+#   @exit()       # internally triggered (more common).
+#
+
+# For debug only; to be stripped out.
+count = 0
+
+class Mode
+  # Static.
+  @modes: []
+
+  # Constants; readable shortcuts for event-handler return values.
+  continueBubbling: true
+  suppressEvent: false
+  stopBubblingAndTrue: handlerStack.stopBubblingAndTrue
+  stopBubblingAndFalse: handlerStack.stopBubblingAndFalse
+
+  # Default values.
+  name: ""
+  badge: ""
+  keydown: null # null will be ignored by handlerStack (so it's a safe default).
+  keypress: null
+  keyup: null
+
+  constructor: (options={}) ->
+    Mode.modes.unshift @
+    extend @, options
+    @modeIsActive = true
+    @count = ++count
+    console.log @count, "create:", @name
+
+    @handlers = []
+    @exitHandlers = []
+
+    @push
+      keydown: @keydown
+      keypress: @keypress
+      keyup: @keyup
+      updateBadge: (badge) => @alwaysContinueBubbling => @chooseBadge badge
+
+    Mode.updateBadge() if @badge
+
+  push: (handlers) ->
+    @handlers.push handlerStack.push handlers
+
+  onExit: (handler) ->
+    @exitHandlers.push handler
+
+  exit: ->
+    if @modeIsActive
+      console.log @count, "exit:", @name
+      handler() for handler in @exitHandlers
+      handlerStack.remove handlerId for handlerId in @handlers
+      Mode.modes = Mode.modes.filter (mode) => mode != @
+      Mode.updateBadge()
+      @modeIsActive = false
+
+  # The badge is chosen by bubbling an "updateBadge" event down the handler stack allowing each mode the
+  # opportunity to choose a badge.  chooseBadge, here, is the default. It is overridden in sub-classes.
+  chooseBadge: (badge) ->
+    badge.badge ||= @badge
+
+  # Shorthand for a long name.
+  alwaysContinueBubbling: (func) -> handlerStack.alwaysContinueBubbling func
```

Comment:
Can you explain the use case for "always continuing to bubble" so the purpose of this mechanism is clear? Either here or in the general comments at the top of mode.coffee